src/Workspaces/Core/Portable/Utilities/BKTree.Builder.cs

```diff
@@ -0,0 +1,286 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Roslyn.Utilities
+{
+    internal partial class BKTree
+    {
+        private class Builder
+        {
+            private const int CompactEdgeAllocationSize = 4;
+
+            private readonly char[] _allLowerCaseCharacters;
```

Comment:
For instance, you use _characterSpans.Length as the word count, which isn't obviously correct from the name here (but I do understand that it is correct)